### DISCUSSION TITLE
getting_started/first_3d_game/06.jump_and_squash

### DISCUSSION THREAD

**Discussion by giscus** [godotengine/godot-docs-user-notes#140] [User-contributed notes]
# getting_started/first_3d_game/06.jump_and_squash

In this part, we'll add the ability to jump and squash the monsters. In the next lesson, we'll make the player die when a monster hits them on the ground. First, we have to change a few settings re...

https://docs.godotengine.org/en/stable/getting_started/first_3d_game/06.jump_and_squash.html

**Comment by vmorris** [comment]
In 4.4, one has to click the plus sign next to "Filter Groups" and then they can create a new group.

**Reply by vmorris** [reply]
Sorry, I meant 4.3.

**Reply by Floflo35** [reply]
And there's a check box to make the group global. I'm wondering if you need to enable it or not.

(edit: you don't need to enable it apparently)

**Comment by vmorris** [comment]
All of my collisions between the mobs and the player have a dot product between 0.4 and 0.6.  .. Not sure why this is.

I changed my evaluation to be < 0.6 to handle this.

**Reply by redisnotbluedev** [reply]
I think that's to do with exactly how the collision shapes are set up, because mine needed to be < 0.7.

**Reply by AriHStein** [reply]
This happens because the player's collision shape is a sphere and the mob's is a rectangle. If the rectangle is not tall enough, the top edge of it will collide with the underside of the sphere causing a collision angle far from 90 degrees. If the rectangle is taller than the sphere (or taller than the widest point on the sphere, really) then when the sphere is on the ground, it will only collide with the rectangle at its widest point, with an angle parallel to the ground plane.

**Reply by reinhillmann** [reply]
This can also happen if a mob spawns when you're busy jumping. That mob will spawn looking at the player and be angled slightly upwards. This causes the dot product to be inaccurate as the mob's box collider would be angled.

**Reply by Calinou** [reply]
- See https://github.com/godotengine/godot-docs-user-notes/discussions/118#discussioncomment-12113928.

**Reply by Orthoclase** [reply]
To add and expand to this: 

The sphere's normal angle at collision can range all the way from just short of an entirely upwards Vector to an entirely downwards Vector (at extremely large and small collision sphere sizes). The resulting dot product between it and the Up Vector can also range between -1.0f and 1.0f. This is only when you choose your collision sphere poorly, but it highlights how the chosen method of using that dot product is fallible.

I found a much easier solution than fiddling with the normal value threshold: just use a cylinder instead of a sphere. The only surface normals of a cylinder are up, down, and sideways vectors.

**Reply by FrackinFamous** [reply]
You can stop the mobs from riding "wheelies" if they are spawned when the player is jumping. We just have to set the x rotation to zero right after we look at the player. 

``` gdscript
func initialize(start_position, player_position) -> void:
    
    look_at_from_position(start_position, player_position, Vector3.UP)
    # Set x rotation to 0 to stop wheelies when player is jumping upon mob instantiation
    rotation.x = 0
```

**Comment by noikoinoi** [comment]
A bit confused about the forLoop in order to detect the collision.
So if we were to detect collision while a player is moving should we always use loops in order to detect collision?
Using loops within physics process/processes seems like bad practice...

**Reply by Calinou** [reply]
Collision detection is an iterative process, so it needs to use a loop. This is common to pretty much all physics implementations in games.

In practice, slide collision count rarely exceeds 4 (and probably never exceeds 8 or so), so it's unlikely to be a performance bottleneck.

Godot already limits the maximum number of physics steps that can run in each frame to avoid "spiral of death" performance scenarios (see the **Max Physics Steps per Frame** project setting).

**Comment by megalon** [comment]
When jumping my first jump was high, the next jump was short, the next was high, and so on.

I fixed this by modifying the jump code to reset the y velocity to zero when on the floor.

```gdscript
if is_on_floor():
	target_velocity.y = 0
	if Input.is_action_just_pressed("jump"):
		target_velocity.y += jump_impulse

if not is_on_floor():
	target_velocity.y -= fall_acceleration * delta
```

**Comment by jclark-dot-org** [comment]
My player bounced on every contact.  Eventually with some `print()` statements to check the generated normals, I settled on bouncing when normal > 0.75.   At least it's working, but I don't really know why.

I'm guessing this is because my colliders aren't quite right on the player and/or mob scenes, but I found those sections of the tutorial to be very hard to follow.  It didn't help that the blue bounding boxes for the collision shapes were much much fainter than in the screen shots, but it also really wasn't clear exactly how they should be set.  Might be useful to have screen shots of the final transforms for all of the objects and their shapes, or something.

**Reply by Sarulean** [reply]
The player collider is a sphere.  It is relatively big (so the player doesn't overlap the mobs) to the box collider on the mob.  The result is mob box collides with the bottom half of the player sphere and you get a collision vector that is downward.  The easiest options are to adjust the sphere to be smaller, make the box collider taller, or increase the threshold for determining downward collision as you did.

**Reply by gold-fx** [reply]
I faced the same problem of the player character bouncing on every contact, and my solution was to set the V Offset of SpawnLocation (the PathFollow3D node) to 1 so that mobs spawned in at the same height as the player model instead of already being below it. I also changed the player's collision shape from a sphere to a cylinder, because I felt it better fit the shape of the model.

**Comment by Mastur1998** [comment]
What does `if collision.get_collider() == null` do? It seems like the condition never checks true.

**Reply by otlol** [reply]
Same question here. I tried:

  if collision.get_collider() == null:
 			print('null happened')
			continue

it only printed when I squashed a mob. 
My guess is that the collision happens but with mob geteting queue_free() and the collider becomes null, hence returning the print(). This does not match with the instruction "# If the collision is with ground". 

When I try print(get_slide_collision_count()), I always get '1' when on floor, so collision with floor is happening whenever player is on the ground. 

Can someone clarify what is the use of "  if collision.get_collider() == null:" part?

**Reply by A-iko** [reply]
This is actually something I ran into myself when trying to run this without the nullcheck.

It seems to me that the physics engine is registering a collision with the collider that SHOULD have been removed after we call queue_free() in the mob. 

I'm not smart enough to tell you if this is expected or acceptable behaviour. It seems rather odd to me.

**Reply by PikminMaster18** [reply]
In the      for index in range(get_slide_collision_count()):   we asked the game to compile a list of all the collisions that occur in a single frame. Now that the game has all the collisions compiled, it needs to know what to do with them. The    
if collision.get_collider() == null:      tells the computer to do nothing with all collisions. The       Continue         that follows tells the game to bring the collisions to the next if statement. The        if collision.get_collider().is_in_group("mob"):          only allows the collisions with nodes in the mob group to follow the next instructions. 
When you ask a script to compile information (an Index compiles information)  it must be used or else the game won't be able to process the script completely which in turn can cause the build to crash.
The       if collision.get_collider() == null:       tells the game to do nothing with the information compiled by the index which allows the game to move on.

**Reply by shimpe** [reply]
In a single frame the player can detect multiple collisions with the same mob. All these collisions are recorded in the list we're iterating over. The first of these collisions causes the script to delete (squash) its collider (i.e. the mob it collided with is deleted and becomes null). By the time we investigate a collision further down the list involving that same (by now already deleted) mob, it will have a collider that is null.

**Comment by Menekrr** [comment]
For me, when trying to adjust the _Collision -> Layer_ and _Collision -> Masks_, the Inspector for ```Player``` and ```Mob``` don't look like that. Specifically:

(1) The _"Collision"_ section underneath _"Moving Platform"_ contains only ```safe_margin```.
(2) There's another section called _"CollisionObject3D"_ slightly lower down, which also contains an item called _"Collision"_. This is the right place to adjust the _Collision -> Layer_ and _Collision -> Masks_.



I'm using a MacBook Pro Retina, Mid-2014:
Godot 4.3 stable
macOS Catalina 10.15.7
CPU: 2.5 GHz Quad-Core Intel Core i7
Memory: 16 GB 1600 MHz DDR3
GPU: NVIDIA GeForce GT 750M 2 GB; Intel Iris Pro 1536 MB

**Reply by knokelmaat** [reply]
This is no bug. I understand why it might be confusing though. A CharacterBody3D inherits PhysicsBody3D, which in turn inherits CollisionObject3D. CharacterBody3D has a property related to collision, namely Safe Margin. A CollisionObject3D has 3 properties related to collision, namely: Layer, Mask and Priority. As `Player` and `Mob` are both CharacterBody3D, they will show both the section for collision related to CharacterBody3D and CollisionObject3D. The `Ground` however, is a StaticBody3D and that one doesn't add any specific properties related to collision, so there is no collision section there, only in the CollisionObject3D that it also inherits.

**Comment by RensTillmann** [comment]
In case someone else happens to run into this issue: My collision against mobs didn't work. Apperantly I accidently "unchecked" the group "mob" I had created without noticing it right after creation. Noticed after I debugged the code and noticed the engine was actually detecting collision just fine based on get_slide_collision() info.

**Comment by LevHol** [comment]
Hi! Maybe someone has dealt with the same issue before.?The collision apparently works, player can even stay on top of the mobs, but they don't get killed. I tried to fix the problem by changing and adjusting CollisionShape3D of the player and the mob, but it only got worse, so I returned to the default set from the tutorial. Any suggestions how to it can be solved?

**Reply by yuexiaohuaz** [reply]
I had the same issue, but it turned out that I forgot to put the closing parentheses at the end of the squash command.

**Comment by RaylaxKai** [comment]
Here's a nice thing you can add to improve the game feel - jump buffering! 

*What's jump buffering?* Have you ever tried to jump a second time the instant you touched the ground, but because you were still *slightly* in the air at the exact moment you hit the button, you didn't get that second jump? That's because right now our logic **only** lets you jump if you're touching the ground at the *moment* you press the button:

```
if is_on_floor() and Input.is_action_just_pressed("jump"):
```

If you're still in the air even by a fraction when you hit the jump button, you won't jump again!

A lot of games improve their *game feel* by including things like *jump buffering* and *coyote time*, which let you jump even if you're veeeeery slightly off with your timing (coyote time is, btw, when a game lets you jump just after you've run off the edge of a ledge, giving you a split second moment to recover rather than plummeting you to your death - like a certain cartoon coyote). And it's easy to implement here!

What we're going to do is, whenever the player presses the jump button, we set a very short timer (not a Godot Timer object in this case, though you could use one of those), a tenth of a second or so long, and count it down to zero. If the player touches the ground and the timer is still above zero, we make them jump. So, if they're already on the ground, they jump immediately, as normal. If they're very slightly off the ground, but touch the ground before the timer hits zero, then they jump as soon as they touch the ground. It gives a little bit of allowance in the player's input when they want to do two jumps in quick succession, and *feels good*

So how to write this? First off, we need two variables at the top of our player.gd script, below `@export var jump_impulse = 20`:

```
@export var jump_buffer = 0.1
var jump_buffer_timer = 0
```

`jump_buffer` is the window of time, in seconds, we give the player to press jump before they touch the ground that we'll "allow" and let them jump again. We'll export this so we can tweak it in the editor - you'll want to play with this value to see what feels good for you, I found 0.1 or 0.2 to work well.
`jump_buffer_timer` is, well, the timer. We'll set this to the value of `jump_buffer` every time we press jump and then count it down every frame until it hits zero.

Next, in `func _physics_process(delta):`, in your block of `if Input.is_action_pressed("...")` commands, add a new one for "jump", which sets the timer:

```
if Input.is_action_just_pressed("jump"):
		jump_buffer_timer = jump_buffer
```

Now we need to change the code we wrote in the tutorial to look at the jump timer rather than the jump input. Change this:

```
if is_on_floor() and Input.is_action_just_pressed("jump"):
		target_velocity.y = jump_impulse
```

to this:

```
if is_on_floor() and jump_buffer_timer > 0:
		target_velocity.y = jump_impulse
		jump_buffer_timer = 0
```

We're looking at the timer now instead of the player input directly - if its greater than 0, that means the player recently hit the jump button, so if they're also touching the floor, they should jump! We also set the timer to 0 - it shouldn't matter, because our jump takes longer than 0.1 seconds, but if you set the buffer to something huge like 5 seconds and don't reset this to 0 when the player does jump, then they're going to keep hopping for the whole five seconds, and we don't want that. So we set it back to zero as a safety measure.

The last thing we need to do is, of course, actually make our countdown timer count down. At the end of the function, add this:

```
if (jump_buffer_timer > 0):
		jump_buffer_timer = clamp(jump_buffer_timer, 0, jump_buffer_timer - delta)
```

This is saying: if the timer is currently greater than zero, reduce it by delta. Clamp the value so if it goes under 0, it gets set to 0. If it's already zero, we don't need do anything.

And that's it! Give it a try! You might find it's a little hard to "notice" the difference if your timer is set to 0.1 or 0.2, so try setting it to something big like 2.0 seconds to test its working. Jump, and then hit the jump button again while you're in midair. As soon as your player touches the ground, they'll jump a second time. Nice! Now set it back to 0.1 or 0.2 and play around with it until it feels right. Congrats, you've implemented a simple jump buffer!

**Reply by Calinou** [reply]
We can probably modify the tutorial to make jumping work continuously when held down (i.e. "pogo stick" behavior). It should only require a minimal amount of code changes to implement, with no added complexity. I find this much preferable in the vast majority of games either way (compared to more limited jump buffering).

**Comment by sirframes** [comment]
Complete Godot newbie here, ran into a peculiar issue, three of the "diagonal directions" would not let the player jump while using the arrow keys. After a quick search I discovered it's actually keyboard ghosting, a hardware related issue, and it works fine using WASD. So watch out for that!
I guess that's why not many games use arrow keys nowadays.

**Comment by ErBimo** [comment]
Squashing not working, how do i check the collisions?

**Reply by ErBimo** [reply]
ok i understood how to check the collisions and the player is not colliding with the mob, help?

**Reply by ErBimo** [reply]
alright i fixed it, i had to activate he mask number 3 for the mob

**Reply by turnerendlesslearning** [reply]
Same issue, same solution. To see why I wasn't getting collisions with mobs, I moved my camera to the side. Looks like Player is hovering slightly above the ground and mobs were slightly inset into the ground. When I checked 3 on mob mask, it raised mobs to same hover as player and allowed collision.

**Reply by ZXSwire3** [reply]
I was having the same issue and checking 3 on the mob masked worked as a solution, but I went back a few steps and it seems like it was an issue with the mob spawn path. My spawn path points were a y = 0 but I think they are intended to be at y = 1. After making this change I was able to remove check 3 on the mob mask and it still worked.

**Comment by acloudmovingby** [comment]
Like @ErBimo, I also strangely could not get it to work unless I turned on mask 3 ("world") for the mob. 

If I didn't, the player would seem to have collision events happen when moving into a mob but it never registered those collisions as with the mob objects. To be specific, if I moved my player into a mob (jumping or not), `get_slide_collision_count()` would go from 1 to 3, but `is_in_group("mob")` always returned false. Once I enabled Mask 3 ("world") for Mob, then it worked as in the tutorial. 

It doesn't make sense to me why this needs to be done and it doesn't match what's described in the tutorial which is a little frustrating

**Reply by ChengHanWu** [reply]
Same thing happened, I just added "print" log and prove that it never detect is_in_group("mob"). I didn't even realized the "world" mask make it work until I saw this post.

**Comment by leviandurai** [comment]
So I'm at the part where you add jumping and I'm getting some odd behaviour. 

First, I have to be moving to jump. It doesn't jump while standing still. Second, once I've jumped the player stays suspended midair if I release the movement keys. It will resume falling when I press the movement keys again. 

```gdscript
extends CharacterBody3D

#How fast the player moves in m/s
@export var speed = 14
# The downward acceleration when in the air, in meters per second squared.
@export var fall_acceleration = 75
# Vertical impulse applied to the character upon jumping in meters per second.
@export var jump_impulse = 20

var target_velocity = Vector3.ZERO 


func _physics_process(delta):
	# We create a local variable to store the input direction.
	var direction = Vector3.ZERO
	# We check for each move input and update the direction accordingly.
	if Input.is_action_pressed("move_right"):
		direction.x += 1
	if Input.is_action_pressed("move_left"):
		direction.x -= 1
	if Input.is_action_pressed("move_back"):
		direction.z += 1
	if Input.is_action_pressed("move_forward"):
		direction.z -= 1
		# Notice how we are working with the vector's x and z axes.
		# In 3D, the XZ plane is the ground plane.
		
	if direction != Vector3.ZERO:
		direction = direction.normalized()
		# Setting the basis property will affect the rotation of the node.
		$Pivot.basis = Basis.looking_at(direction)
		
		# Ground Velocity
		target_velocity.x = direction.x * speed
		target_velocity.z = direction.z * speed
		
		#Vertical Velocity
		if not is_on_floor(): #aka if in the air
			target_velocity.y = target_velocity.y - (fall_acceleration * delta)
		if is_on_floor() and Input.is_action_just_pressed("jump"):
			target_velocity.y = jump_impulse
			
		# Moving the Character
		velocity = target_velocity
		
		move_and_slide()
```

**Reply by GradXL49** [reply]
Hello! The bug in your code is that a large portion is mistakenly in the "if direction" block. From your comment "# Ground Velocity" down, that all needs to be tabbed back once to move out of the above if statement. Hope that helps!

**Comment by umarthdc** [comment]
My character barely moves when falling
When printing it seems that multiplying (75 * delta) gives a really small number like 0.33. 
Is this correct?

**Comment by MehmetEmreAksu** [comment]
my player does not jump when i was going top left or bottom right. 

```gdscript
extends CharacterBody3D

@export var speed = 14
@export var fall_acceleration = 75
@export var jump_impulse = 20
@export var bounce_impulse = 16

var target_velocity = Vector3.ZERO


func _physics_process(delta):
	var direction = Vector3.ZERO

	if Input.is_action_pressed("move_right"):
		direction.x += 1
	if Input.is_action_pressed("move_left"):
		direction.x -= 1
	if Input.is_action_pressed("move_back"):
		direction.z += 1
	if Input.is_action_pressed("move_forward"):
		direction.z -= 1

	if direction != Vector3.ZERO:
		direction = direction.normalized()
		# Setting the basis property will affect the rotation of the node.
		$Pivot.basis = Basis.looking_at(direction)

	# Ground Velocity
	target_velocity.x = direction.x * speed
	target_velocity.z = direction.z * speed

	# Vertical Velocity
	if not is_on_floor(): # If in the air, fall towards the floor. Literally gravity
		target_velocity.y = target_velocity.y - (fall_acceleration * delta)

	# Moving the Character

	if is_on_floor() and Input.is_action_just_pressed("jump"):
		target_velocity.y = jump_impulse
	# Iterate through all collisions that occurred this frame
	for index in range(get_slide_collision_count()):
		# We get one of the collisions with the player
		var collision = get_slide_collision(index)

		# If there are duplicate collisions with a mob in a single frame
		# the mob will be deleted after the first collision, and a second call to
		# get_collider will return null, leading to a null pointer when calling
		# collision.get_collider().is_in_group("mob").
		# This block of code prevents processing duplicate collisions.
		if collision.get_collider() == null:
			continue

		# If the collider is with a mob
		if collision.get_collider().is_in_group("mob"):
			var mob = collision.get_collider()
			# we check that we are hitting it from above.
			if Vector3.UP.dot(collision.get_normal()) > 0.1:
				# If so, we squash it and bounce.
				mob.squash()
				target_velocity.y = bounce_impulse
				# Prevent further duplicate calls.
				break
				
	velocity = target_velocity
	move_and_slide()
```
	
this is the code

**Reply by jumpthe11** [reply]
Tried your code everything works right. Maybe your keyboard has a ghosting? Try changing input map keys to different ones.

**Reply by MehmetEmreAksu** [reply]
aynen aga ghosting varmış wasd ile çalıştı ok tuşlarıyla olmuyodu. yine de sağolasın

**Comment by acidhills** [comment]
To prevent squashing mobs at the beginning of a jump, you can add velocity check like 
```
_targetVelocity.Y < 0 
```

**Comment by MrPaolinsky** [comment]
Something really cool you can achieve with the godot typing system is getting autocompletion on "mob" and also making sure whatever returns collision.get_collider() is actually a mob.

First inside mob.gd at the top of the file after "extends CharacterBody3D" add the following:
```GDScript
class_name Mob
```
```GDscript
var mob := collision.get_collider() as Mob 

if not mob: # Mob will be null if the cast on the previous line fails
   continue

mob.squash() # Here you should get autocompletion on your code editor.
```

If you want to learn more about type casts, and typing in general you can check: https://docs.godotengine.org/en/stable/tutorials/scripting/gdscript/static_typing.html

**Reply by knokelmaat** [reply]
Very cool stuff, thanks for adding this!

**Reply by lakejeetech-blip** [reply]
Was wondering if GD script had some kind of type system!

**Reply by GithubNanior** [reply]
Yea exactly! Best part is no need to specially create groups = Simpler, less prone to setup mistakes and faster too, in all manners the better way... Although I get that they use groups here to demonstrate the feature.
(Also didn't know that null nodes evaluate to false here, that's like Unity, I like it)

**Comment by shangyuq** [comment]
Within the codeblock that checks if we are landing on the monster (*if Vector3.UP.dot(collision.get_normal()) > 0.1:*),
I used *print(mob)* to inspect the instances being hit.
The console gaved me 2 kind of output when player hit the mob:
1.Mob:<CharacterBody3D#139536106980>
2.@CharacterBody3D@23:<CharacterBody3D#185941886413>
I wonder why there is a second instance? Shouldn't be all *Mob*? Thank you.

**Comment by Pinaz993** [comment]
I hesitated a long time on this chapter, because I wasn't sure where the `for` loop was supposed to go in  `_process_physics(delta)`. I finally realized that it had to happen after `move_and_slide()` because it depended on data gathered during the execution of that method. Maybe make that a little clearer by adding `move_and_slide()` to the code block right after the `#...` and/or explicitly stating it in the instructions above the code block.

In addition, if the player happens to be in mid air when a mob spawns, the mob will be pointed at the player's elevated position, instead of flying level. To solve that, I added `rotation[0] = 0` to `Mob.initialize()` right after the random rotation is calculated. Perhaps something to add to this tutorial?

**Comment by focusgg** [comment]
To those of you who have found that after completing this section that your player technically squashes all enemies even when just bumping into them, make your player's CollisionShape3D shape a CylinderShape3D. An immediate fix without having to find what dot product nonsense would work for the placement of your Player's sphere.

**Comment by SzyVsSi** [comment]
Since we're checking for collisions here anyway, what are the arguments against using `move_and_collide()` for the player character? Newbie asking :)

**Comment by fait-5** [comment]
This might be nitpicking a little too much but, since the Player can jump now, the mobs sometimes rotate on their x axis making them appear as if they were looking a little bit upwards, which seems weird sometimes because that makes part of their body clip through the ground a little bit. I easily fixed this by just adding a line to the initialize function

```gdscript
func initialize(start_position: Vector3, player_position: Vector3) -> void:
	look_at_from_position(start_position, player_position, Vector3.UP)
	rotation.x = 0
	rotate_y(randf_range(-PI / 4, PI / 4))
	
	var random_speed := randi_range(min_speed, max_speed)
	velocity = random_speed * Vector3.FORWARD
	velocity = velocity.rotated(Vector3.UP, rotation.y)
```

**Comment by clawrence54** [comment]
Hello! The player can move and jump just fine, and the mobs spawn and move. However, when I jump on top of a mob, it gets pushed through the ground without dying. I've copied and pasted the code checks and it still happens. I also checked the collision masks. Any help appreciated!

**Comment by admin** [comment]
*Originally posted by [Soulhae](https://github.com/Soulhae):*

> i got confused with the collision loop, why is the if statement before move_and_slide()? if i'm not mistaken, get_slide_collision_count() returns collision count from the last call to move_and_slide(). so if I follow what the docs say, wouldn't the logic be 1 frame behind?

**Reply by admin** [reply]
*Originally posted by [Khamler](https://github.com/Khamler):*

> No, the order of events is as follows:
>
> -move_and_slide detects a collision
> -next physics frame starts
> -the collision loop finds the collision and decides if a bounce impulse should occur on the next move_and_slide
> -move_and_slide is called again and the bounce impulse is applied

**Reply by admin** [reply]
*Originally posted by [Khamler](https://github.com/Khamler):*

> In other words, since move_and_slide() is what takes bounce impulse, the bounce impulse will always be applied the physics tick after the collision is detected (unless you called move_and_slide() twice per physics tick).
>
> Therefore, your version has the same outcome as their version, except that the bounce impulse is calculated before all the other changes in velocity, which would be bad because gravity would be calculated after the bounce impulse.
